Add optional async polling of button states

Ricket opened the pull request:
(in progress -- builds & tested; see todo list below)

Today Controllable checks button states in the client tick, which appears to be tied to the framerate. So if you have a framerate limit set to 30 or 40 fps, then it's easy to rapidly press a button where Controllable won't notice that you released & pressed it. For example, if you rapidly press R1, the number of hotbar slot changes is often less than the number of button presses. You can repro this _very_ reliably by reducing framerate limit as low as 10 or 20.

Per discussion in #344 , GLFW doesn't support callback-based gamepad button events, so we have to poll. Which means to fix this issue, we need a background thread that can poll at a higher rate; 125Hz is a common polling rate for peripherals so I think we should aim for at least that.

I started by refactoring the controller button state polling, separate from the button handling. Then added a toggle for whether to do the polling async (in its own thread) or to do it in the client tick like today. If async, then a ScheduledExecutorService is used to trigger the polling at a fixed rate. The polling side is only GLFW calls, nothing touching Minecraft at all, so it should be ok on a background thread.

The poll makes a new `ButtonStates`, compares it to the previous and adds any new button presses/releases to a thread-safe queue of button state events. Then, the client tick handler consumes the entries in the queue and handles each one as a button state change.

Test plan:
* Video settings > Framerate limit set to 10 or 20.
* Rapid press L1 or R1, see if it always moves the correct number of hotbar slots.
* Rapid press swap hands button, see that it always swaps the correct number of times
* Rapid press camera change button, see that it always changes camera
* I expect inventory navigation should feel more responsive too
* Button binding should still work
* Radial menu should still work

Todo:

- [ ] @MrCrayfish should the default for this feature be on or off? Or should we only do async polling and not even give the option?
- [x] Add config GUI setting for polling frequency, and move the async enabled into the config GUI too.
- [x] Thoroughly test to make sure there's no issues.

Out of scope for this PR, a follow-up change could be made to separate out the events that don't need to wait for the client tick. For example, buttons that move the mouse around, should be able to be processed immediately (or at least queued for immediate execution on the main thread). So that those can benefit from 125Hz or the video frame rate, and not be stuck processing these at 20Hz. Or the events which have to send a packet to the server and wait for the response, can kick that packet off sooner and thus feel quicker.

Fixes #344

Ricket committed:
Add optional async polling of button states

MrCrayfish commented:
The problem is that GLFW calls must only be invoked on the main thread. Idealy I should add an option to swtich input to be processed on RenderTickEvent instead of ClientTickEvent, as that will have a higher polling rate assuming framerate is higher than 20. The issue is also present on the Bedrock edition.

Ricket commented:
Hm, yeah I see it in the glfw and lwjgl docs, all of the glfw calls being used say they need to run from the main thread. Then we can just wrap the glfw calls in `Minecraft.getInstance().runImmediately`, right? I'll experiment with it soon. It's funny that this worked in my testing but I thought I felt a couple rare hiccups, maybe those were related. Also the exit crashed and now that I think about it, that was probably due to this, or the thread making glfw calls after it was shutting down. I thought it was just a fluke, but actually this probably needs a safer thread stop...

I don't think tying it to RenderTickEvent would give better responsiveness; apparently the ClientTickEvent is tied to render speed, since repro is as easy as setting framerate limit. And sure, framerate limit is probably not 20 in normal usage, but 30 or 40 fps is reasonable, and I initially noticed it very quickly with 40 fps.

Ricket committed:
Do the glfw operations in the main thread

Ricket committed:
Do not poll when Minecraft is shutting down

Ricket commented:
`runImmediately`, ironically, does not run immediately but on average waits for the next frame. However, `enqueue` does tend to run immediately because it wakes the thread. I found that by using `Minecraft.getInstance().enqueue()`, I can do all the glfw stuff on the main thread like it's supposed to, but also the result is still more responsive. It's not as good at 10fps, still misses some inputs... but 30fps and 40fps are still a lot more responsive than without the async poller thread. 30fps misses an occasional input, and 40fps I couldn't get it to miss any inputs. Much better than processing it on client tick.

I added assertions in the methods that are making the glfw calls, just to be extra sure that they are only happening on the main thread, like the docs say it needs to be.

I also added a check to make sure Minecraft is running, so that the poller stops doing anything when the game is shutting down.

Ricket committed:
Add asyncPolling toggle to settings page

Ricket commented:
I tested with higher frequencies and I think 125Hz is sufficient, in fact 1000Hz actually performed worse, so I think we only need an enabled toggle, and I think the background thread should always run at 125Hz.

I added the toggle to the settings page.

@MrCrayfish do you want this default on or off? Maybe you can try it out and see for yourself, if you agree it's useful. I would love to see it default on, but I also understand if you hesitate and want to default it off. Either way I will turn it on for my own computers and steam deck...

Otherwise, I think this is ready for any more testing you want to do and then ready to merge. I've only played around with it for a few minutes but I don't see any issues. Note that this is off the 1.16 branch so will need to be forward-ported to the other versions...

![image](https://user-images.githubusercontent.com/196978/201282141-15cbbc83-6935-4560-9680-e871a5317ef3.png)

MrCrayfish commented:
I think there is a misunderstanding of how the game works. Since the gamepad must be polled on the main thread only, scheduling tasks from a seperate thread is no different than just having a simple for loop executed on `RenderTickEvent`, an event that is run every game loop just like executing enqueued tasks. You would also not need the overhead of an extra thread.
```
> start game loop
> run enqueued tasks
> tick game if enough time elsaped
> render frame (render tick event)
> end game loop
```
Assuming FPS is limited to 10 at a 125Hz poll rate, ~12 tasks will be equeued every frame. How realistical is it for the gamepad state to change inbetween the execution of those tasks, certainly too quick for anyone to physically change the button. There would need to be a big time difference between the execution of each task.

Also another problem I found is that input is being executed on the next frame when async is off. This is due to `enqueue` call.

The only solution I can see is checking the gamepad state before and after a game tick and before and after rendering a frame since they are big time chunks in the game loop.
```
> start game loop
> run enqueued tasks
    * check gamepad state, queue for processing, run queue
> tick game if enough time elsaped
    * check gamepad state and queue for processing
> render frame (render tick event)
    * check gamepad state and queue for processing
> end game loop
```

Ricket commented:
Changing `enqueue` to `execute` fixes the issue of input executed next frame when async off.

But yeah, I see the bigger problem. I logged the time between polls and it's like you said, several polls happen right after another. The main thread is still managing to sleep for 33ms, which is right because I have the FPS limit set to 30 fps. But with async off, it's tied to the client tick, which is every 50ms so it's alternating sleeping 33ms and 66ms.

![image](https://user-images.githubusercontent.com/196978/201389652-e4624035-904a-4539-9875-5069c0b89571.png)

So this solution is no better than tying it to the RenderTickEvent. I was misled because `enqueue` calls `LockSupport.unpark(this.getExecutionThread());` which I assumed/hoped was waking the main thread early. Instead, the way Minecraft sleeps, looks like it calls `glfwWaitEventsTimeout` in a loop until the desired amount of sleep elapses.

![image](https://user-images.githubusercontent.com/196978/201391343-1f9bad0e-eab0-45cb-bbaf-4bb51bb8c495.png)

glfw only has `glfwPostEmptyEvent`, which would just wake it up to promptly go right back to sleep. I guess there's no allowed way to poll the controller in the background, then.

Seems like [this PR](https://github.com/glfw/glfw/pull/1590) would give us what we want, but it stalled for the past few years, although there was activity in the last month...

So, yeah I agree with you, it seems like the `ButtonStateTracker` stuff in this PR could still come in handy to implement the solution of checking the gamepad state before & after each big chunk of the main loop, and bundling/queueing up the changes for the next client tick. We can't avoid the between-frames sleep but at least we can poll a few times while we're awake.

Ricket committed:
Fix 1-frame lag when async is off

MrCrayfish commented:
I feel as `ButtonStateTracker` is redundant. All you need to do is push `ButtonStates` created each time polling the gamepad to the original `Controllable#processButtonStates` method and it will handle tracking the state.

MrCrayfish commented:
I've gone ahead and implemented my solution to fix the original issue. A controller should now be more responsive if your FPS is low (due to performance). If your FPS is high but you decide to cap to 40 or lower, I've added an option in the config that allows the controller to be polled while the game is applying the artifical fps limit.

Feel free to check it out. I am still testing it, mainly with other mods like Optifine/Rubidium since they might do some weird stuff.

Ricket commented:
Any chance you could backport your fix to the 1.16.x branch and cut a release? We're still playing Enigmatica 6 Expert and I was just updating my Steam Deck, noticed there hasn't been an update since August.

Also, closing this PR, thanks for the better fix.
